[hdfs] Support host and port in HDFS URIs

vvaten opened the pull request:
Fixes #162.

The original PR by @vvaten correctly identified that smart_open's `ParseUri` mangled HDFS URIs in a way that's against [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986). Since 2018, HDFS parsing moved out of `smart_open_lib.py` into `smart_open/hdfs.py`, but the underlying bug came along for the ride: `parse_uri` still collapses `netloc + path` into a single slash-prefixed string, so `hdfs://host:8020/tmp/test.txt` becomes `/host:8020/tmp/test.txt` before reaching the `hdfs dfs` CLI — and the CLI no longer addresses the right cluster.

This PR ports the original intent against the current code: when `netloc` is non-empty, `parse_uri` returns the **full URI** so the CLI can resolve the cluster itself.

### Before vs after

| Input URI                         | `parse_uri(...)['uri_path']` (before) | `parse_uri(...)['uri_path']` (after) | Effective `hdfs dfs` call            |
|-----------------------------------|----------------------------------------|---------------------------------------|--------------------------------------|
| `hdfs:///tmp/test.txt`            | `/tmp/test.txt`                        | `/tmp/test.txt` *(unchanged)*         | default cluster, `/tmp/test.txt`     |
| `hdfs://host/tmp/test.txt`        | `/host/tmp/test.txt` 🐛                | `hdfs://host/tmp/test.txt`            | `host` cluster, `/tmp/test.txt`      |
| `hdfs://host:8020/tmp/test.txt`   | `/host:8020/tmp/test.txt` 🐛           | `hdfs://host:8020/tmp/test.txt`       | `host:8020` cluster, `/tmp/test.txt` |
| `viewfs://cluster/tmp/test.txt`   | `/cluster/tmp/test.txt` 🐛             | `viewfs://cluster/tmp/test.txt`       | `cluster` mount table resolves       |

`URI_EXAMPLES` is updated to reflect the supported shapes (the misleading `hdfs://path/file` form is removed in favor of `hdfs://host/path/file` and `hdfs://host:port/path/file`).

### Docs

- `help.txt` regenerated via `python update_helptext.py`.
- `README.rst`: replaced `hdfs://path/file` and `hdfs://user/hadoop/my_file.txt` with explicit `hdfs://host:port/...` forms.

### Tests

- New parametrized `test_parse_uri` / `test_parse_uri_invalid` in `tests/test_hdfs.py` covering the three netloc variants for `hdfs`/`viewfs` plus error cases.
- Existing `tests/test_smart_open.py::*::test_hdfs` mocks updated to assert the full-URI passthrough for both read and write paths.

Local run: `pytest tests/test_hdfs.py tests/test_smart_open.py -k hdfs` → 32 passed.

Commit:
added netloc support for hdfs URIs; removed the special handling of HDFS URIs which was against the URI specification

mpenkov commented:
Thank you for your pull request. It looks good to me.

@menshikh-iv Might be a good idea to merge this _after_ our HDFS integration tests are up. What do you think?

menshikh-iv commented:
@mpenkov I agree.
@vvaten thanks for PR! Sorry for waiting, but first we need to finish https://github.com/RaRe-Technologies/smart_open/issues/151.

menshikh-iv commented on smart_open/smart_open_lib.py:
```diff
@@ -355,6 +356,9 @@ class ParseUri(object):
       * file:///home/user/file
       * file:///home/user/file.bz2
 
+    NOTE: hdfs://path/file does no longer work as it is against the URI
```
Can you add more information to comment (when this happens, what HDFS version affected, etc)?

vvaten replied:
This is a generic change that affects all HDFS versions. They do support hdfs://host/path/file URI format. Using hdfs://path/file in smart_open violates this and also violates the URI specification (RFC3986) where the hostname part is always after the first two / characters. The correct way to refer to local content is hdfs:///path/file instead of hdfs://path/file.

govindmurthi21 commented:
Hello is there any way this can be merged ? As part of my work I was trying to read and write to hdfs and traced the bug to this 
Hdfs commands require a fully qualified uri such as ‘’hdfs dfs ls hdfs://path’’

ddelange commented:
Hi @vvaten :wave:

I know it's been a while. It looks like this PR is still valid, only outdated. Do you want to pull `develop` and get this merged?

ddelange committed:
Merge remote-tracking branch 'origin/develop' into pr-168-hdfs-netloc

* origin/develop: (797 commits)
  [azure] Mark Writer/AppendWriter closed even when close() raises (#924)
  [http,azure] Optimize forward seeks within buffered data to avoid redundant requests (#923)
  Update CHANGELOG.md
  Add support for lz4 compression (#921)
  Update CHANGELOG.md
  [hdfs] Make CliRawOutputBase.write return number of bytes written (#920)
  Bump the github-actions group with 2 updates (#919)
  Update CHANGELOG.md
  Add support for `append` mode for Azure Blob Storage (#865)
  Bump softprops/action-gh-release in the github-actions group (#918)
  Update CHANGELOG.md
  Fix pylance complaint (#917)
  Add more references to README.rst (#915)
  Pin github-release-commenter (#914)
  Fix doctest (#913)
  Bump softprops/action-gh-release in the github-actions group (#912)
  Bump minimum supported version to Python 3.10 (#911)
  Bump the github-actions group with 2 updates (#910)
  Update CHANGELOG.md
  [s3] Support very large buckets in iter_bucket, add client_kwars (#908)
  ...

# Conflicts:
#	smart_open/smart_open_lib.py
#	smart_open/tests/test_smart_open.py

ddelange committed:
[hdfs] Preserve full URI when netloc is present

The original PR (#168) by @vvaten ported smart_open_lib's ParseUri to
preserve the full HDFS URI so the hdfs CLI can route to the right
cluster.  Since then, HDFS parsing moved to smart_open/hdfs.py but the
underlying bug remained: parse_uri collapses netloc+path into a slash-
prefixed local-looking path, so "hdfs://host:8020/tmp/test.txt" becomes
"/host:8020/tmp/test.txt" and the CLI no longer addresses the cluster.

This commit ports the original intent against the current code: if a
netloc is present, parse_uri returns the full URI as uri_path; otherwise
it returns the absolute path (for "hdfs:///path" forms).  URI_EXAMPLES is
updated to reflect the supported shapes, and the test suites cover the
three netloc variants for both read and write paths.

Fixes #162.

ddelange committed:
[hdfs] Update docs to match parse_uri changes

- Regenerate help.txt from the new URI_EXAMPLES via update_helptext.py.
- README.rst: replace misleading "hdfs://path/file" and "hdfs://user/..."
  examples with explicit "hdfs://host:port/..." forms, matching how the
  CLI now resolves the cluster.
